File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> res;
    void solve(vector<int>& nums, int target, int ind, vector<int> &temp){
        if(target == 0){
            res.push_back(temp);
            return;
        }

        if(target < 0 || ind >= nums.size())
            return;
        

        for(int i = ind; i < nums.size(); i++){
            // handle duplicate elements
            if(i > ind && nums[i] == nums[i-1])
                continue;
            // include current element
            if(nums[i] <= target){
                temp.push_back(nums[i]);
                solve(nums, target - nums[i], i + 1, temp);
                temp.pop_back();
            }          
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> temp;
        sort(candidates.begin(), candidates.end());
        solve(candidates, target, 0, temp);
        return res;
        
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new ArrayList<>();
        solve(candidates, target, 0, new ArrayList<>(), ans);
        return ans;
    }

    public void solve(int[] candidates, int target, int start, List<Integer> temp, List<List<Integer>> ans){

        if(target == 0){
            ans.add(new ArrayList<>(temp));
            return;
        }

        for(int i = start; i < candidates.length; i++){
            // if(candidates[i] <= target){
            //     if(i > start && candidates[i] == candidates[i - 1])
            //         continue;
            //     temp.add(candidates[i]);
            //     solve(candidates, target - candidates[i], i + 1, temp, ans);
            //     temp.remove(temp.size() - 1);
            // }

            if(candidates[i] > target)
                break;
            
            if(i > start && candidates[i] == candidates[i - 1])
                    continue;

            temp.add(candidates[i]);
            solve(candidates, target - candidates[i], i + 1, temp, ans);
            temp.remove(temp.size() - 1);
        }
    }
}